zebROS_ws/src/tf_object_detection/src/object_detection.cpp

```diff
@@ -26,30 +29,288 @@ void camera_info_callback(const sensor_msgs::CameraInfoConstPtr &info)
 	caminfovalid = true;
 }
 
-// Get the average of the values in the cv::Mat depth contained within
-// the supplied bounding rectangle
-double avgOfDepthMat(const cv::Mat& depth, const cv::Rect& bound_rect)
+// Find the median of a continuous cv::Mat
+float findMedianOfMat(cv::Mat mat) {
+	float median = 0;
+	if (mat.isContinuous()) {
+		// copy matrix data to a vector
+		std::vector<float> vec;
+		mat = mat.reshape(0, 1);
+		mat.copyTo(vec);
+		// remove 255 (when this is called, values that are masked are 255)
+		vec.erase(std::remove(vec.begin(), vec.end(), 255), vec.end());
+		// sort vector
+		std::sort(vec.begin(), vec.end()); // forgot this earlier
+		if ((vec.size() % 2) != 0) { // if odd
+			auto it = vec.begin() + vec.size()/2;
+			std::nth_element(vec.begin(), it, vec.end());
+			median = vec[vec.size()/2];
+		} else { // if even
+			auto it = vec.begin() + vec.size()/2;
+			std::nth_element(vec.begin(), it, vec.end());
+			median += vec[vec.size()/2];
+			it = vec.begin() + vec.size()/2 - 1;
+			std::nth_element(vec.begin(), it, vec.end());
+			median += vec[vec.size()/2];
+			median = ((float)median/2.0d);
+		}
+	}
+	return median;
+}
+
+// K_MEANS = k-means, Forgy method of initializing (k-means++ would likely be better, but Forgy is working ok for now)
+// CONTOURS = find depth by finding contours and taking the lowest value of the biggest contour
+// CONTOURS_NON_ADAPTIVE = contours, but using thresholding based on the median instead of adaptive thresholding
+enum DepthCalculationAlgorithm { K_MEANS, CONTOURS, CONTOURS_NON_ADAPTIVE };
+
+// Get the most useful depth value in the cv::Mat depth contained within
+// the supplied bounding rectangle, using contour finding
+double contoursDepthMat(const cv::Mat& depth_, const cv::Rect& bound_rect, bool debug = false, bool adaptive = true) {
+	if (bound_rect.size().area() == 0) { // if the ROI is zero, return -1 (no depth)
+		return -1;
+	}
+
+	// Crop depth to region of interest
+	cv::Mat depth = depth_(bound_rect);
+
+	// set very large outliers and nan to 0 so they can be removed later
```

Comment:
I really like this code (I honestly didn't know about setTo, sounds like something which could be GPU optimized in the future, heh).  We might want add a TODO to see if the zed actually reports negative depth - if not, could remove a few of these operations to speed things up later.

Reply:
Ok, thanks! I added the TODO and hopefully we'll be in person soon to resolve it :)